//   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto2";

package evo_kit;

message EvoKitConfig {
  //sampling configuration
  optional int32 seed = 1 [default = 18];
  optional int32 buffer_size = 2 [default = 100000];
  optional GaussianSamplingConfig gaussian_sampling = 3;
  // Optimizer Configuration
  optional OptimizerConfig optimizer = 4;
  // AsyncESAgent Configuration
  optional AsyncESConfig async_es = 5;
}

message GaussianSamplingConfig {
  optional float std = 1 [default = 1.0];
  optional bool cached = 2 [default = false];
  optional int32 cache_size = 3 [default = 100000];
}

message OptimizerConfig{
  optional string type = 1 [default = "SGD"];
  optional float base_lr = 2 [default = 1e-3]; // The base learning rate.
  optional float momentum = 3 [default = 0.9]; // The momentum value for SGD.

  // ------------Adam Optimizer---------
  optional float beta1 = 4 [default = 0.9];
  optional float beta2 = 5 [default = 0.999];
  optional float epsilon = 6 [default = 1e-8];
}

message SamplingInfo{
  repeated int32 key = 1;
  optional int32 model_iter_id = 2;
}

message AsyncESConfig{
  optional string model_warehouse = 1 [default = "./model_warehouse"];
  repeated string model_md5 = 2;
  optional int32 max_to_keep = 3 [default = 5];
  optional int32 model_iter_id = 4 [default = 0];
}
